### PR TITLE
refactor: merge `http.errors` with `api.response`

### PR DESCRIPTION
**Pull request by SnickerDoodlePop**

### CODE REVIEW

**Comment by SnickerDoodlePop** [comment]
Docstrings are NOT yet written.

File: virtool/api/response.py [hunk 1]
```diff
@@ -26,27 +36,73 @@ def json_response(
...
         self.errors = errors
+
+
+@web.middleware
+async def middleware(req: web.Request, handler: Callable):
+    """
+    INCOMPLETE FUNCTION DOCSTRING
+
+    :param req:
+    :param handler:
```
Comment (ReeceHoffmann) [inline virtool/api/response.py:97]:
```suggestion
    Request handling middleware for converting HTTP exceptions into HTTP responses

    :param req: the incoming request
    :param handler: the next middleware to be executed
```

**Review by ReeceHoffmann — CHANGES_REQUESTED** [review]
Code restructure looks good, added some docstring suggestions. Feel free to stray from the suggestions. Write additional docstrings for the custom exceptions

**Review by ReeceHoffmann — CHANGES_REQUESTED** [review]

File: virtool/api/response.py [hunk 2]
```diff
@@ -1,8 +1,19 @@
+"""
+Virtool API Response
+
+Exports HTTP Errors and request handling middleware for reformatting and reporting
+errors to the client.
```
Comment (igboyes) [inline virtool/api/response.py:5]:
```suggestion
HTTP error exceptions and middleware for reformatting and reporting
errors to the client.
```

File: virtool/api/response.py [hunk 3]
```diff
@@ -26,27 +37,72 @@ def json_response(
 
 
 class InsufficientRights(HTTPForbidden):
+    """
+    Reusable exception for returning an HTTP status `403` (Insufficient Rights) message
+    to the client.
```
Comment (igboyes) [inline virtool/api/response.py:42]: * How is it used?
  > Raising this exception during request handling immediately returns a `403` (Insufficient Rights) response to the client.
* "Reusable" doesn't have much meaning here.
* Please use this wording for the other docstrings.

File: virtool/api/response.py [hunk 4]
```diff
@@ -26,27 +37,72 @@ def json_response(
...
+    client.
+    """
+
     def __init__(self, errors, message="Invalid input"):
         super().__init__(text=message, reason="invalid_input")
         self.errors = errors
+
+
+@web.middleware
+async def middleware(req: web.Request, handler: Callable):
```
Comment (igboyes) [inline virtool/api/response.py:92]: Use `from` imports where this is imported.
```suggestion
async def error_middleware(req: web.Request, handler: Callable):
```

File: virtool/api/response.py [hunk 5]
```diff
@@ -26,27 +37,72 @@ def json_response(
...
+    """
+    Request handling middleware for converting HTTP exceptions into HTTP responses.
+
+    :param req: the incoming request
+    :param handler: the next middleware to be executed
+    """
+    try:
+        resp = await handler(req)
+        return resp
```
Comment (igboyes) [inline virtool/api/response.py:101]:
```suggestion
        return await handler(req)
```

File: virtool/api/response.py [hunk 6]
```diff
@@ -26,27 +37,72 @@ def json_response(
...
+
     def __init__(self, errors, message="Invalid input"):
         super().__init__(text=message, reason="invalid_input")
         self.errors = errors
+
+
+@web.middleware
+async def middleware(req: web.Request, handler: Callable):
+    """
+    Request handling middleware for converting HTTP exceptions into HTTP responses.
```
Comment (igboyes) [inline virtool/api/response.py:94]: There is no type of middleware other than "request handling" in the application.
```suggestion
    Middleware for converting HTTP exceptions into HTTP responses.
    
    Catches any `HTTPException` raised during request handling and returns a response with the appropriate status code and JSON body.
```

**Review by igboyes — CHANGES_REQUESTED** [review]

**Review by igboyes — APPROVED** [review]

**Pull request merged**